File: 트리/leetcode783(Easy)-BST 노드 간 최소 거리.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# DFS, 재귀구조로 중위순회 // 16ms (90%)
import sys
class Solution(object):
    
    prev, result = -sys.maxsize, sys.maxsize
    
    def minDiffInBST(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root.left:
            self.minDiffInBST(root.left)

        self.result = min(self.result, root.val - self.prev)
        self.prev = root.val

        if root.right:
            self.minDiffInBST(root.right)

        return self.result


# DFS, 반복구조로 중위순회
import sys
class Solution(object):
    def minDiffInBST(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        prev, result = -sys.maxsize, sys.maxsize
        
        stack, node = [], root
        
        while stack or node:
            while node:
                stack.append(node)
                node = node.left
                
            node = stack.pop()
            
            result = min(result, node.val - prev)
            prev = node.val
            
            node = node.right
        
        return result

File: 트리/leetcode783(Easy)-BST 노드 간 최소 거리.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

// 중위순회(재귀) : 64ms(78.82%), 42.1MB(96.47%)
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDiffInBST = function (root) {
  let prev = -Infinity
  let result = Infinity

  const inOrder = (node) => {
    if (node.left) {
      inOrder(node.left)
    }

    result = Math.min(result, node.val - prev)
    prev = node.val

    if (node.right) {
      inOrder(node.right)
    }
  }

  inOrder(root)

  return result
}

// 중위순회(스택) : 66ms(71.18%), 43.1MB(62.35%)
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDiffInBST = function (root) {
  let prev = -Infinity
  let result = Infinity

  let node = root
  const stack = []
  while (stack.length > 0 || node) {
    while (node) {
      stack.push(node)
      node = node.left
    }

    node = stack.pop()
    result = Math.min(result, node.val - prev)
    prev = node.val

    node = node.right
  }

  return result
}
